integration-tests/test/test_dag_correctness.py

```diff
@@ -1,19 +1,113 @@
+                        deploy1.start()
+
+                        deploy2 = DeployThread("validator2", validator2, contract_path, 10)
+                        deploy2.start()
+
+                        deploy3 = DeployThread("validator3", validator3, contract_path, 10)
+                        deploy3.start()
+
+
+                        deploy1.join()
+                        deploy2.join()
+                        deploy3.join()
+
+                        expected_blocks_count = 31
+                        wait_for_blocks_count_at_least(
+                            context,
+                            validator1,
+                            expected_blocks_count,
+                        )
+                        wait_for_blocks_count_at_least(
+                            context,
+                            validator2,
+                            expected_blocks_count,
+                        )
+                        wait_for_blocks_count_at_least(
+                            context,
+                            validator3,
+                            expected_blocks_count,
+                        )
+
+                        with bootstrap_connected_peer(context=context, bootstrap=bootstrap_node, name='bonded-validator-4', keypair=BONDED_VALIDATOR_KEY_4) as validator4:
+
+                            deploy4 = DeployThread("catch_up", validator1, contract_path, 1)
+                            deploy4.start()
+                            deploy4.join()
+
+                            wait_for_blocks_count_at_least(
+                                context,
+                                validator1,
+                                expected_blocks_count + 1,
+                            )
+                            wait_for_blocks_count_at_least(
+                                context,
+                                validator2,
+                                expected_blocks_count + 1,
+                            )
+                            wait_for_blocks_count_at_least(
+                                context,
+                                validator3,
+                                expected_blocks_count + 1,
+                            )
+                            wait_for_blocks_count_at_least(
+                                context,
+                                validator4,
+                                expected_blocks_count + 1,
+                            )
+
+                            validator1_vdag = validator1.get_vdag()
+                            validator2_vdag = validator2.get_vdag()
+                            validator3_vdag = validator3.get_vdag()
+                            validator4_vdag = validator4.get_vdag()
+
+                            assert validator1_vdag == validator4_vdag
+                            assert validator2_vdag == validator4_vdag
+                            assert validator3_vdag == validator4_vdag
```

Comment:
If any blocks get orphaned, it is possible that these vdags will look different.

Reply:
Do you have any recommended way to compare two dag?
I tried to use networkx to parse the string but the finalized block info would be lost.

Reply:
show-blocks should give information about finalized blocks. If you could just check that the main chain's are the same that would be enough.

Reply:
Wait a minute. 
If any blocks get orphaned, this test should fail as expected.
Should a dag with an orphaned block be considered the same as a dag without orphaned blocks?

BTW, how to create a dag with orphaned blocks? I can not produce this kind of test case in my own machine.

Reply:
Yeah if you remove all the orphaned blocks from both dags, then they should be the same.

It is difficult to create an orphaning when you got good network connectivity.

Reply:
Nvm, I realized my mistake.

If a validator A has proposed a block A1 followed by block A2 and A1 is orphaned, then the resulting dag for some validators may not include A1. However, if A2 is included in all validator's dag's, then since A1 is in the justification of A2 (as you always point to your own previous block in your justification), A1 will automatically be pulled in if A2 is included.
The only case in which A1 won't exist on another validator's dag is if A2 also is orphaned (along with all the blocks above it by validator A). Hence, the only time the dags will be different is if the network continues to be down forever for a specific validator - which I think you don't have to worry about in this test.

Reply:
I've expressed my opinion on how I see it should work at https://rchain.atlassian.net/browse/RCHAIN-2925
@rabbitonweb what's the progress on that ticket?

Reply:
@adaszko My suggestion here is that we merge this pr first since this test case won't cause an orphaned branch or something. We can deal with the more complicated feature on another pr(ticket).